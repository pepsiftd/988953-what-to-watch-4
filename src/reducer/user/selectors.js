import {NameSpace} from '@/reducer/name-space';
import {AuthorizationError} from '@/reducer/user/user';

const getAuthorizationStatus = (state) => {
  return state[NameSpace.USER].authorizationStatus;
};

const getBadRequestStatus = (state) => {
  return state[NameSpace.USER].authorizationError === AuthorizationError.BAD_REQUEST;
};

export {getAuthorizationStatus, getBadRequestStatus};
