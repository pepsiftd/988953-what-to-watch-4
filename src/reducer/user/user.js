import {extend} from '@/utils';
import {UserInfoModel} from '@/models/user-info-model';

const AuthorizationStatus = {
  AUTHORIZED: `AUTHORIZED`,
  UNAUTHORIZED: `UNAUTHORIZED`,
};

const AuthorizationError = {
  NO_ERROR: 200,
  BAD_REQUEST: 400,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.AUTHORIZED,
  authorizationError: AuthorizationError.NO_ERROR,
  authorizationInfo: {},
};

const ActionType = {
  SET_AUTHORIZATION_STATUS: `SET_AUTHORIZATION_STATUS`,
  SET_AUTHORIZATION_ERROR: `SET_AUTHORIZATION_ERROR`,
  SET_AUTHORIZATION_INFO: `SET_AUTHORIZATION_INFO`,
};

const ActionCreator = {
  setAuthorizationStatus: (status) => {
    return {
      type: ActionType.SET_AUTHORIZATION_STATUS,
      payload: status,
    };
  },
  setAuthorizationError: (status) => {
    return {
      type: ActionType.SET_AUTHORIZATION_ERROR,
      payload: status,
    };
  },
  setAuthorizationInfo: (authorizationInfo) => {
    return {
      type: ActionType.SET_AUTHORIZATION_INFO,
      payload: authorizationInfo,
    };
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_AUTHORIZATION_STATUS:
      return extend(state, {
        authorizationStatus: action.payload,
      });
    case ActionType.SET_AUTHORIZATION_ERROR:
      return extend(state, {
        authorizationError: action.payload,
      });
    case ActionType.SET_AUTHORIZATION_INFO:
      return extend(state, {
        authorizationInfo: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.setAuthorizationStatus(AuthorizationStatus.AUTHORIZED));
        dispatch(ActionCreator.setAuthorizationInfo(UserInfoModel.parseUserInfo(response.data)));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authorizationData, onSuccess = () => {}, onFail = () => {}) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authorizationData.email,
      password: authorizationData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.setAuthorizationStatus(AuthorizationStatus.AUTHORIZED));
        dispatch(ActionCreator.setAuthorizationError(AuthorizationError.NO_ERROR));
        dispatch(ActionCreator.setAuthorizationInfo(UserInfoModel.parseUserInfo(response.data)));
        onSuccess();
      })
      .catch((err) => {
        dispatch(ActionCreator.setAuthorizationError(err.response.status));
        onFail(err);
      });
  },
};

export {initialState, ActionType, ActionCreator, Operation, reducer, AuthorizationStatus, AuthorizationError};
