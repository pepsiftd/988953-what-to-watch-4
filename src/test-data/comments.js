const reviews = [
  {
    id: 1,
    user: {
      id: 4,
      name: `Courtney Cox`,
    },
    rating: 5.1,
    comment: `I don't know what to say but I can't help posting`,
    date: new Date(`August 3, 2020 UTC`),
  },
  {
    id: 2,
    user: {
      id: 2,
      name: `Mockingbird`,
    },
    rating: 8.3,
    comment: `This is not a movie of a simple kind`,
    date: new Date(`July 15, 2020 UTC`),
  },
  {
    id: 3,
    user: {
      id: 154,
      name: `Pelagia`,
    },
    rating: 1.5,
    comment: `Super mega exquisitely boring`,
    date: new Date(`June 1, 2020 UTC`),
  }
];

export {reviews};
