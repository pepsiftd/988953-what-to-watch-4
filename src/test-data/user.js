const userInfo = {
  id: 0,
  email: `example@email.com`,
  name: `Sarah Connor`,
  avatar: `img/avatar.jpg`,
};

export {userInfo};
