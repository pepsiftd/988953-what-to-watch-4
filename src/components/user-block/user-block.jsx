import React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import {AuthorizationStatus} from '@/reducer/user/user';
import {AppRoute} from '@/const';


const UserBlock = ({authorizationStatus, avatarImageSrc = `img/avatar.jpg`}) => {
  return (
    <div className="user-block">
      {authorizationStatus === AuthorizationStatus.AUTHORIZED &&
        <div className="user-block__avatar">
          <Link to={AppRoute.MY_LIST}>
            <img src={avatarImageSrc} alt="User avatar" width="63" height="63" />
          </Link>
        </div>
      }
      {authorizationStatus === AuthorizationStatus.UNAUTHORIZED &&
        <Link to={AppRoute.LOGIN} className="user-block__link">Sign in</Link>
      }

    </div>
  );
};

UserBlock.propTypes = {
  authorizationStatus: PropTypes.oneOf(Object.values(AuthorizationStatus)).isRequired,
  avatarImageSrc: PropTypes.string,
};

export {UserBlock};
