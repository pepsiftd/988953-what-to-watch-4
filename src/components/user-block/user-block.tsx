import React from 'react';
import {Link} from 'react-router-dom';
import {AuthorizationStatus} from '@/reducer/user/user';
import {AppRoute} from '@/const';
import {AuthInfo} from '@/types';

interface Props {
  authorizationStatus: AuthorizationStatus;
  authorizationInfo: AuthInfo;
}


const UserBlock: React.FunctionComponent<Props> = ({authorizationStatus, authorizationInfo}) => {
  const {avatar} = authorizationInfo;
  return (
    <div className="user-block">
      {authorizationStatus === AuthorizationStatus.AUTHORIZED &&
        <div className="user-block__avatar">
          <Link to={AppRoute.MY_LIST}>
            <img src={avatar} alt="User avatar" width="63" height="63" />
          </Link>
        </div>
      }
      {authorizationStatus === AuthorizationStatus.UNAUTHORIZED &&
        <Link to={AppRoute.LOGIN} className="user-block__link">Sign in</Link>
      }

    </div>
  );
};

export {UserBlock};
