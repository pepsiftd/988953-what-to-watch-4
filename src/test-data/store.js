import {NameSpace} from '@/reducer/name-space';
import {movies, promoMovie} from '@/test-data/movies';
import {userInfo} from '@/test-data/user';
import {reviews} from '@/test-data/comments';
import {ALL_GENRES_FILTER} from '@/const';

const store = {
  [NameSpace.DATA]: {
    movies,
    promoMovie,
    favoriteMovies: [],
    reviews,
  },
  [NameSpace.APP]: {
    currentGenre: ALL_GENRES_FILTER,
  },
  [NameSpace.USER]: {
    authorizationStatus: `AUTHORIZED`,
    authorizationError: 200,
    authorizationInfo: userInfo,
  },
};

export {store};
