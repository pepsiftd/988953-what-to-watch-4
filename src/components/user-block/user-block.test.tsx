import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Router} from 'react-router-dom';
import {history} from '@/history';
import {userInfo} from '@/test-data/user';

import {UserBlock} from './user-block';


describe(`UserBlock`, () => {
  it(`should render a login link when unauthorized`, () => {
    const tree = renderer.create(
        <Router history={history}>
          <UserBlock
            authorizationStatus="UNAUTHORIZED"
            authorizationInfo={{}}
          />
        </Router>
    );

    expect(tree).toMatchSnapshot();
  });

  it(`should render avatar image when authorized`, () => {
    const tree = renderer.create(
        <Router history={history}>
          <UserBlock
            authorizationStatus="AUTHORIZED"
            authorizationInfo={userInfo}
          />
        </Router>
    );

    expect(tree).toMatchSnapshot();
  });
});
