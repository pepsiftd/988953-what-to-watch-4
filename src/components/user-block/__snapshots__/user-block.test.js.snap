// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserBlock should render a login link when unauthorized 1`] = `
<div
  className="user-block"
>
  <a
    className="user-block__link"
    href="/login"
    onClick={[Function]}
  >
    Sign in
  </a>
</div>
`;

exports[`UserBlock should render avatar image when authorized 1`] = `
<div
  className="user-block"
>
  <div
    className="user-block__avatar"
  >
    <a
      href="/mylist"
      onClick={[Function]}
    >
      <img
        alt="User avatar"
        height="63"
        src="img/avatar.jpg"
        width="63"
      />
    </a>
  </div>
</div>
`;
